api/apps/geoprocessing/src/modules/surface-cost/application/surface-cost-processor.ts

```diff
@@ -6,25 +6,30 @@ import { CostSurfaceJobInput } from '../cost-surface-job-input';
 
 import { CostSurfacePersistencePort } from '../ports/persistence/cost-surface-persistence.port';
 import { PuExtractorPort } from '../ports/pu-extractor/pu-extractor.port';
-import { ArePuidsAllowedPort } from '../ports/pu-validator/are-puuids-allowed.port';
+import { GetAvailablePlanningUnits } from '../ports/available-planning-units/get-available-planning-units';
 import { ShapefileConverterPort } from '../ports/shapefile-converter/shapefile-converter.port';
+import { lockPlanningUnits } from '@marxan/scenarios-planning-unit';
 
 @Injectable()
 export class SurfaceCostProcessor
   implements WorkerProcessor<CostSurfaceJobInput, true> {
   constructor(
     private readonly repo: CostSurfacePersistencePort,
     private readonly puExtractor: PuExtractorPort,
-    private readonly puValidator: ArePuidsAllowedPort,
+    private readonly availablePlanningUnits: GetAvailablePlanningUnits,
     private readonly shapefileConverter: ShapefileConverterPort,
   ) {}
 
   async process(job: Job<CostSurfaceJobInput, true>): Promise<true> {
     const geoJson = await this.shapefileConverter.convert(job.data.shapefile);
     const surfaceCosts = this.puExtractor.extract(geoJson);
-    const { errors } = await this.puValidator.validate(
-      job.data.scenarioId,
+    const planningUnitsIds = (
+      await this.availablePlanningUnits.get(job.data.scenarioId)
+    ).ids;
+
+    const { errors } = lockPlanningUnits(
       surfaceCosts.map((cost) => cost.planningUnitId),
+      planningUnitsIds,
     );
     if (errors.length > 0) {
       throw new Error(errors.join('.'));
```

Comment:
given the way we use these errors, and referencing my earlier comment about this, I think we could even return the ids of "unknown" scenario-planning-units and generate here the error message to tell the user which ones are not part of the scenario - we leak abstractions a bit, but at the end this is still a `CostSurfaceProcessor` (or `SurfaceCostProcessor`? 😅), and I think we can assume we know about the concept of scenario-planning-units in this class.

Reply:
The message already does contain those IDs - and the data (of failure) is not (for now on) processed in any manner 🤔

Reply:
yep, we discussed this (in a similar context? can't remember the details) - just saying that if we concatenate errors here, we may as well also only return the naked ids from `canPlanningUnitsBeLocked()` and do the templating here instead of returning full error messages, but I am happy with your latest updates in `canPlanningUnitsBeLocked()`.